### PR TITLE
Add osx-arm64 + switch to rattler-build

### PR DESCRIPTION
**Pull request by hadim**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by hadim** [comment]
@conda-forge-admin, please rerender

**Comment by hadim** [comment]
The build passes locally on a MacBook Pro M3 but fails here. I guess it's due to the cross compiling.

**Comment by hadim** [comment]
@conda-forge/xformers I am happy to keep digging here. I tried to look around for this `_build_env/lib/python3.11/site-packages/torch/lib/libtorch_global_deps.dylib' (no such file)` error but could not find anything.

I feel like there is some confusion in between the host and build envs. Also worth noting, the build passes locally on `osx-arm64`.

Any idea? Did I miss something obvious here?

**Comment by hadim** [comment]
A new bug that seems to be due to rattler this time:

```bash
 × error Error building package: Failed to resolve dependencies: Cannot solve the request because of: No candidates were found for __osx >=11.0.
Error:   × Failed to resolve dependencies: Cannot solve the request because of: No
  │ candidates were found for __osx >=11.0.
  │ 
  ╰─▶ Cannot solve the request because of: No candidates were found for __osx
      >=11.0.
```

**Comment by hadim** [comment]
Seems related to https://github.com/prefix-dev/rattler-build/issues/1052. I wonder whether fixing this would also fix the above cross compilation issue using `meta.yaml`.

**Comment by hadim** [comment]
(and yes, using rattler decreases the build duration from ~15 min to ~5 min on osx-64)

**Comment by hadim** [comment]
Looking at https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1025920&view=logs&jobId=1b8be447-c2bd-5772-b66a-a1146441bf88&j=1b8be447-c2bd-5772-b66a-a1146441bf88&t=0da7bde9-21e8-5d19-f418-738876121ca7, it seems that `target_platform` is set `osx-64` instead of `osx-arm64` for an `osx-arm64` build. Could that explain the virtual package error?

**Comment by hadim** [comment]
```
+ rattler-build build --recipe ./recipe -m ./.ci_support/osx_arm64_python3.11.____cpython.yaml --output-dir /Users/runner/miniforge3/conda-bld --no-test

 ╭─ Finding outputs from recipe
 │ Found 1 variants
 │ Build variant: xformers-0.0.27-cpu_py311h1f6929c_1
 │ 
 │ ╭───────────────────────┬─────────────────────────────────────────────────────────────────────────────────────────────────────────╮
 │ │ Variant               ┆ Version                                                                                                 │
 │ ╞═══════════════════════╪═════════════════════════════════════════════════════════════════════════════════════════════════════════╡
 │ │ CONDA_BUILD_SYSROOT   ┆ /Applications/Xcode_14.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk │
 │ │ build_platform        ┆ osx-64                                                                                                  │
 │ │ c_stdlib              ┆ macosx_deployment_target                                                                                │
 │ │ c_stdlib_version      ┆ 11.0                                                                                                    │
 │ │ channel_targets       ┆ conda-forge main                                                                                        │
 │ │ cuda_compiler         ┆ None                                                                                                    │
 │ │ cuda_compiler_version ┆ None                                                                                                    │
 │ │ cxx_compiler          ┆ clangxx                                                                                                 │
 │ │ cxx_compiler_version  ┆ 17                                                                                                      │
 │ │ python                ┆ 3.11.* *_cpython                                                                                        │
 │ │ pytorch               ┆ 2.3                                                                                                     │
 │ │ target_platform       ┆ osx-64                                                                                                  │
 │ ╰───────────────────────┴─────────────────────────────────────────────────────────────────────────────────────────────────────────╯
 │
 ╰─────────────────── (took 0 seconds)
```

**Comment by hadim** [comment]
@conda-forge-admin, please rerender

**Comment by hadim** [comment]
osx works but osxarm64 fails and it seems related to https://github.com/prefix-dev/rattler-build-conda-compat/issues/56

```
  File "/Users/runner/miniforge3/lib/python3.12/site-packages/conda_forge_ci_setup/feedstock_outputs.py", line 158, in main
    distributions = built_distributions_from_recipe_variant(recipe_dir=recipe_dir, variant=variant)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/runner/miniforge3/lib/python3.12/site-packages/conda_forge_ci_setup/utils.py", line 96, in built_distributions_from_recipe_variant
    allowed_dist_names, allowed_subdirs = get_built_distribution_names_and_subdirs(
                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/runner/miniforge3/lib/python3.12/site-packages/joblib/memory.py", line 577, in __call__
    return self._cached_call(args, kwargs, shelving=False)[0]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/runner/miniforge3/lib/python3.12/site-packages/joblib/memory.py", line 532, in _cached_call
    return self._call(call_id, args, kwargs, shelving)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/runner/miniforge3/lib/python3.12/site-packages/joblib/memory.py", line 771, in _call
    output = self.func(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/runner/miniforge3/lib/python3.12/site-packages/conda_forge_ci_setup/utils.py", line 55, in get_built_distribution_names_and_subdirs
    metas = rattler_build_conda_compat.render.render(
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/runner/miniforge3/lib/python3.12/site-packages/rattler_build_conda_compat/render.py", line 314, in render
    m.config.variant = package_variants[0]
                       ~~~~~~~~~~~~~~~~^^^
IndexError: list index out of range
```

**Comment by hadim** [comment]
Blocked by https://github.com/prefix-dev/rattler-build-conda-compat/issues/56

**Comment by hadim** [comment]
It seems like https://github.com/conda-forge/conda-smithy/pull/2070#discussion_r1782970359 remove a couple of issues here but it still fails with:

```
  File "/Users/runner/miniforge3/lib/python3.12/site-packages/rattler_build_conda_compat/render.py", line 314, in render
    m.config.variant = package_variants[0]
```

Note that while the build platform is set to `osx-64` during the build phase looking at the logs, it seems like the build platform is set to `osx-arm64` during the validation step:

```
 ╭─ Finding outputs from recipe
 │ Found 1 variants
 │ Build variant: xformers-0.0.28.post1-cpu_py39h7fd2b17_0
 │ 
 │ ╭───────────────────────┬─────────────────────────────────────────────────────────────────────────────────────────────────────────╮
 │ │ Variant               ┆ Version                                                                                                 │
 │ ╞═══════════════════════╪═════════════════════════════════════════════════════════════════════════════════════════════════════════╡
 │ │ CONDA_BUILD_SYSROOT   ┆ /Applications/Xcode_14.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk │
 │ │ build_platform        ┆ osx-arm64                                                                                               │
 │ │ c_stdlib              ┆ macosx_deployment_target                                                                                │
 │ │ c_stdlib_version      ┆ 11.0                                                                                                    │
 │ │ channel_targets       ┆ conda-forge main                                                                                        │
 │ │ cuda_compiler         ┆ None                                                                                                    │
 │ │ cuda_compiler_version ┆ None                                                                                                    │
 │ │ cxx_compiler          ┆ clangxx                                                                                                 │
 │ │ cxx_compiler_version  ┆ 17                                                                                                      │
 │ │ python                ┆ 3.9.* *_cpython                                                                                         │
 │ │ pytorch               ┆ 2.4                                                                                                     │
 │ │ target_platform       ┆ osx-arm64                                                                                               │
 │ ╰───────────────────────┴─────────────────────────────────────────────────────────────────────────────────────────────────────────╯

```

@wolfv any idea what is going on here?

**Comment by wolfv** [comment]
Hey @hadim no idea! There should be a conda-smithy release soon with teh fix for `--target-platform`. There was also a rattler-build build (number 1) that fixes the `__osx` detection issue. 

With those two fixes, things should look decent for osx-arm64.

**Comment by hadim** [comment]
This PR already contains the `--target-platform` fix and the last CI is using the new rattler-build.

**Comment by wolfv** [comment]
Cool! I'll have to take a look tomorrow. Loooks like something going wrong for validation for osx-arm64!

**Comment by hadim** [comment]
**Current status**

- `linux-64`: **OK**
- `osx-64`: **OK**
- `linux-64` with Cuda: **NOT OK**

I can't get `${{ env.get("CUDA_COMPILER_VERSION", default="None")}}` to correctly resolve to the cuda version, it's always `None`. I also tried using `cuda_compiler_version` directly without success.

Note that the only way to set the build string is using `pytorch [version='*', build='cuda*']  and `pytorch =*=cuda*` does not seem to work.

- `osx-arm64`: **NOT OK**

It works locally (no cross compilation) but fails on the CI (likely because of cross compilation).

The current error is during the validation step: `m.config.variant = package_variants[0]`. See https://github.com/prefix-dev/rattler-build-conda-compat/issues/58

**Comment by wolfv** [comment]
@hadim thank you for relentlessly testing. This is amazing. I wasn't even aware that the syntax you posted works, to be honest :)

I thought that setting the build string would also work with the following syntax: `foo 1.2.* bla*`. If this syntax doesn't work then IMO it's a bug.

**Comment by hadim** [comment]
> Another bug to investigate: when forcing cuda with `cuda_version: ${{ env.get("CUDA_COMPILER_VERSION", default="12.0")}}`, the build fails to install cuda because of missing `__cuda` during the docker build.

Might be related to https://github.com/prefix-dev/rattler-build/pull/1063 right?

**Comment by wolfv** [comment]
@hadim I think we need to double chekc two things:

- Is the `CUDA_COMPILER_VERSION` variable definitely set?
- Does conda-build work by adding `CONDA_OVERRIDE_CUDA` for the `__cuda` availability?

Sorry that this is (not yet) a smooth experience, by the way. It's super valuable that you are trying. 

For `__cuda` the issue you pointed out would indeed be the correct one. I believe that since the latest release `CONDA_OVERRIDE_CUDA` would also work as an env var to "fake" the cuda virtual package.

**Comment by wolfv** [comment]
Btw for most of the things you are trying it should be super quick debuggable locally with rattler-build by doing something like 

```
rattler-build build --recipe ./recipe -m ./.ci_config/bla.yaml --render-only --target-platform=...
```

That should only take a few milliseconds to render out the recipe (and if you also want to solve, then you can add `--with-solve`).

**Comment by hadim** [comment]
> Btw for most of the things you are trying it should be super quick debuggable locally with rattler-build by doing something like
> 
> ```
> rattler-build build --recipe ./recipe -m ./.ci_config/bla.yaml --render-only --target-platform=...
> ```
> 
> That should only take a few milliseconds to render out the recipe (and if you also want to solve, then you can add `--with-solve`).

Using it already and it's super useful indeed :-)

**Comment by hadim** [comment]
> Is the CUDA_COMPILER_VERSION variable definitely set?

Yes it is according to the logs. And if I set it manually and use `rattler-build`, it is also not set.

**Comment by wolfv** [comment]
Let me take a look at the env var problem right now!

**Comment by hadim** [comment]
Yes, sorry, this works fine (just tested locally as well). What does not seem to work is when building with `bash .scripts/run_docker_build.sh`.

There, just before the `conda info` I can indeed see `CUDA_COMPILER_VERSION=12.0` in the logs, but it is not propagated to the `rattler-build` command.

Also I just noticed that `conda info` shows `__cuda=12.0=0` in the docker build so it seems like that one is not propagated to rattler as well.

**Comment by wolfv** [comment]
Alright! Do you have a link to the logs? They seem to all be waiting for the GPU runner right now

**Comment by hadim** [comment]
I am testing locally (no access to the gpu runner) and it's faster anyway.

Here are the logs where I exit just after the `rattler-build` command: [logs.txt](https://github.com/user-attachments/files/17232995/logs.txt)

**Comment by hadim** [comment]
Also with 

```
$ echo $CONFIG
linux_64_cuda_compilercuda-nvcccuda_compiler_version12.0cxx_compiler_version12python3.10.____cpython
```

**Comment by hadim** [comment]
Also the correct variable to use seems to be `CUDA_VERSION`, this is what I am using locally in the recipe file (see https://github.com/conda-forge/conda-forge-ci-setup-feedstock/blob/367229158acb6873003058534ccfdc03976f8ec4/recipe/run_conda_forge_build_setup_linux#L82C1-L82C13).

**Comment by hadim** [comment]
Ok so `CUDA_VERSION` is actually not expoerted while `CONDA_OVERRIDE_CUDA` is. By using `CONDA_OVERRIDE_CUDA` in the recipe it's all good. One problem solved :-)

Next one is the missing `__cuda` virtual package missing.

**Comment by hadim** [comment]
> * Does conda-build work by adding `CONDA_OVERRIDE_CUDA` for the `__cuda` availability?

Yes this is how it is done for conda-build and it seems like it's not supported for rattler-build but according to the doc it seems to be supported by pixi right?

**Comment by wolfv** [comment]
I will double check that this works in rattler-build! :)

**Comment by hadim** [comment]
By install and running `pixi info` in `build_step.sh`, cuda is correctly detected:

```bash
+ /home/conda/.pixi/bin/pixi info
      Pixi version: 0.30.0
          Platform: linux-64
  Virtual packages: __unix=0=0
                  : __linux=6.10.4=0
                  : __glibc=2.17=0
                  : __cuda=12.0=0
                  : __archspec=1=x86_64
         Cache dir: /tmp/.cache/rattler/cache
      Auth storage: /home/conda/.rattler/credentials.json
  Config locations: No config files found
```

but looking at the `rattler-build` command in `build_step.sh`, cuda is not there:

```
Resolving build environment:
 │ │   Platform: linux-64 [__unix=0=0, __linux=6.10.4=0, __glibc=2.17=0, __archspec=1=x86_64]
```

Maybe I am missing something obvious here...

**Comment by wolfv** [comment]
It's probably a bug then! Will try to fix it tonight or tomorrow :)

**Comment by hadim** [comment]
> It's probably a bug then! Will try to fix it tonight or tomorrow :)

As a side note, maybe worth introducing `rattler-build info` while fixing it if it's not too much additional work?

**Comment by wolfv** [comment]
Good idea, should be doable! :)

**Comment by wolfv** [comment]
@hadim this is the fix https://github.com/prefix-dev/rattler-build/pull/1081 to make the CONDA_OVERRIDE_XXX work.

I want to add a `system-requirements` field at some point to the recipe so that they can define these explicitly and simplify cross-compilation and lower bounds on these virtual packages. 

But until then! Let's make it work :)

**Comment by hadim** [comment]
I confirm the `__cuda` and virtual packages override will be fixed once https://github.com/prefix-dev/rattler-build/pull/1081 is released. 

The only remaining issue will be the cross compilation build for `osx-arm64`. That one is tricky to get since I only have an `osx-arm64` local machine.

**Comment by wolfv** [comment]
@hadim I am preparing a 0.23.0 release that should hit the shelves tonight / tomorrow morning. If you find other issues with rattler-build, let me know! :)

**Comment by wolfv** [comment]
Release 0.23.0 is baking: https://github.com/conda-forge/rattler-build-feedstock/pull/89 :)

**Comment by hadim** [comment]
Thanks, @wolfv, for helping here!

We might not be that far, and it will likely help other feedstocks switch to rattler-build in the future.

---

## Current status

- `osx-64`: ✅
- `linux-64-cpu`: ✅ (only locally)
- `linux-64-cuda120`: ✅ (only locally)
- `linux-64-cuda118`: ✅ (only locally)
- `osx-arm64`: ❌ (but works locally without cross compilation)

As said above already, `osx-arm64` fails when it's being cross compiled from `osx-64` in the CI. It's quite tricky for me to debug it and I don't quite understand what it's going on there.

## Next steps

- @h-vetinari: do you want to push a dummy commit here just to trigger the Linux CI and see what happens? 🙏
- If anyone has an idea about the `osx-arm64` cross compilation failure, you're welcome! 🆘

**Comment by wolfv** [comment]
@hadim there are two failure modes for osx-arm64:

- validation fails for 3 of the builds _after_ it built correctly
- on one of the builds, things fail because of a duplicate key (3.12). Maybe some selector is wrong?

I'll try to take a look into the validation asap :)

**Comment by hadim** [comment]
> validation fails for 3 of the builds after it built correctly

yes the builds are all good indeed. thank you for looking into it.

> on one of the builds, things fail because of a duplicate key (3.12). Maybe some selector is wrong?

I thought it was a transient issue or a cache error and ignored it until now. Will have a look.

**Comment by hadim** [comment]
That was simply a leftover `CONDA_BUILD_SYSROOT` key from a previous build. It's gone now.

**Comment by hadim** [comment]
Some linux Cuda builds got green light on Azure CI and the failing ones are only due to lack of disk space.

So it looks like linux builds (cuda or not) are all good here ✅.

**Comment by hadim** [comment]
I think I found the issue. The build and target platform were not correctly propagated during validation.

See the two PRs below to make it work:

- https://github.com/conda-forge/conda-forge-ci-setup-feedstock/pull/360
- https://github.com/prefix-dev/rattler-build-conda-compat/pull/59

**Comment by hadim** [comment]
@conda-forge-admin, please rerender.

**Comment by hadim** [comment]
@h-vetinari LGTM

**Comment by hadim** [comment]
test

**Comment by hadim** [comment]
The linux cuda and non-cuda builds pass locally on my machine.

@h-vetinari this is ready for review.

Do you want me to add myself to https://github.com/conda-forge/.cirun/blob/master/.access.yml and https://github.com/Quansight/open-gpu-server/blob/main/access/conda-forge-users.json or maybe you can trigger it yourself by pushing an empty commit here?

**Review by h-vetinari — APPROVED** [review]
Basically LGTM, thanks. You can of course request to be added to the open-gpu-server, if you'd like (I don't decide, but it makes sense IMO).

File: recipe/recipe.yaml [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+          - libcufft-dev
+          - libcurand-dev
+          - libcusolver-dev
+          - libcusparse-dev
+
+tests:
+  - python:
+      imports:
+        - xformers
+      pip_check: true
```
[the remark this answers is not in the document]
Comment (hadim) [reply recipe/recipe.yaml:118]: Yes, it is, but especially for newcomers, I like to make this explicit.

Ping here if you prefer me to remove it.

[the remark this answers is not in the document]
Comment (h-vetinari) [reply recipe/recipe.yaml]: No, it's OK to leave, your rationale is fine

**Comment by h-vetinari** [comment]
Sorry, for the pileup. I thought I'd fix up your commit so we get a CI run going

**Comment by hadim** [comment]
Ah ok all good.

I let you take over here from now!

**Comment by h-vetinari** [comment]
Actually, I think we can try switching this back to azure. The recent runs on the server only took ~30 min (not hours like it used to; for reasons I don't really understand BTW).

**Comment by hadim** [comment]
> Actually, I think we can try switching this back to azure. The recent runs on the server only took ~30 min (not hours like it used to; for reasons I don't really understand BTW).

I am fine with that since last time I tried, it worked fine.

You do it or I do it?

**Comment by h-vetinari** [comment]
> You do it or I do it?

Sorry, was already in the thick of it. ;-)

**Comment by hadim** [comment]
I bet the osx-arm64 is a transient error. I think a simple restart should do the job.

**Comment by h-vetinari** [comment]
> I bet the osx-arm64 is a transient error. I think a simple restart should do the job.

I would have just restarted the build. Since it's one of the fast ones, that would have taken less time (<10min) than now restarting the linux+CUDA jobs that had already run for ~20min.

**Comment by hadim** [comment]
Yup sorry for that and the error happens again so not sure what is going on actually:

```
thread 'tokio-runtime-worker' panicked at /Users/runner/miniforge3/conda-bld/rattler-build_1734085054975/_build_env/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.42.0/src/runtime/time/entry.rs:568:9:
A Tokio 1.x context was found, but it is being shutdown.
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'tokio-runtime-worker' panicked at /Users/runner/miniforge3/conda-bld/rattler-build_1734085054975/_build_env/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.42.0/src/runtime/time/entry.rs:568:9:
A Tokio 1.x context was found, but it is being shutdown.
thread 'tokio-runtime-worker' panicked at /Users/runner/miniforge3/conda-bld/rattler-build_1734085054975/_build_env/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.42.0/src/runtime/time/entry.rs:568:9:
A Tokio 1.x context was found, but it is being shutdown.
thread 'tokio-runtime-worker' panicked at /Users/runner/miniforge3/conda-bld/rattler-build_1734085054975/_build_env/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.42.0/src/runtime/time/entry.rs:568:9:
A Tokio 1.x context was found, but it is being shutdown.
Error: 
  × Could not collect run exports
  ├─▶ an io error occurred: failed to unpack `/Users/runner/Library/Caches/
  │   rattler/cache/pkgs/pytorch-2.5.1-cpu_mkl_py312h2233c75_106/lib/
  │   python3.12/site-packages/torch/_inductor/runtime/triton_heuristics.py`
  ├─▶ failed to unpack `/Users/runner/Library/Caches/rattler/cache/pkgs/
  │   pytorch-2.5.1-cpu_mkl_py312h2233c75_106/lib/python3.12/site-packages/
  │   torch/_inductor/runtime/triton_heuristics.py`
  ├─▶ failed to unpack `lib/python3.12/site-packages/torch/_inductor/runtime/
  │   triton_heuristics.py` into `/Users/runner/Library/Caches/rattler/cache/
  │   pkgs/pytorch-2.5.1-cpu_mkl_py312h2233c75_106/lib/python3.12/site-
  │   packages/torch/_inductor/runtime/triton_heuristics.py`
  ├─▶ error decoding response body
  ├─▶ request or response body error
  ├─▶ error reading a body from connection
  ╰─▶ stream error received: unexpected internal error encountered

```

**Comment by h-vetinari** [comment]
And it didn't even help, because you still have a transient error now (they seem to be common for some reason, see also the [logs](https://github.com/conda-forge/xformers-feedstock/actions/runs/12380780118) from the run on the GPU server).

In any case, please rebase out the last commit

**Comment by hadim** [comment]
> Yup sorry for that and the error happens again so not sure what is going on actually:
> 
> ```
> thread 'tokio-runtime-worker' panicked at /Users/runner/miniforge3/conda-bld/rattler-build_1734085054975/_build_env/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.42.0/src/runtime/time/entry.rs:568:9:
> A Tokio 1.x context was found, but it is being shutdown.
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> thread 'tokio-runtime-worker' panicked at /Users/runner/miniforge3/conda-bld/rattler-build_1734085054975/_build_env/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.42.0/src/runtime/time/entry.rs:568:9:
> A Tokio 1.x context was found, but it is being shutdown.
> thread 'tokio-runtime-worker' panicked at /Users/runner/miniforge3/conda-bld/rattler-build_1734085054975/_build_env/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.42.0/src/runtime/time/entry.rs:568:9:
> A Tokio 1.x context was found, but it is being shutdown.
> thread 'tokio-runtime-worker' panicked at /Users/runner/miniforge3/conda-bld/rattler-build_1734085054975/_build_env/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.42.0/src/runtime/time/entry.rs:568:9:
> A Tokio 1.x context was found, but it is being shutdown.
> Error: 
>   × Could not collect run exports
>   ├─▶ an io error occurred: failed to unpack `/Users/runner/Library/Caches/
>   │   rattler/cache/pkgs/pytorch-2.5.1-cpu_mkl_py312h2233c75_106/lib/
>   │   python3.12/site-packages/torch/_inductor/runtime/triton_heuristics.py`
>   ├─▶ failed to unpack `/Users/runner/Library/Caches/rattler/cache/pkgs/
>   │   pytorch-2.5.1-cpu_mkl_py312h2233c75_106/lib/python3.12/site-packages/
>   │   torch/_inductor/runtime/triton_heuristics.py`
>   ├─▶ failed to unpack `lib/python3.12/site-packages/torch/_inductor/runtime/
>   │   triton_heuristics.py` into `/Users/runner/Library/Caches/rattler/cache/
>   │   pkgs/pytorch-2.5.1-cpu_mkl_py312h2233c75_106/lib/python3.12/site-
>   │   packages/torch/_inductor/runtime/triton_heuristics.py`
>   ├─▶ error decoding response body
>   ├─▶ request or response body error
>   ├─▶ error reading a body from connection
>   ╰─▶ stream error received: unexpected internal error encountered
> ```

@wolfv any idea what could be causing this? I thought it was transient, but it seems not.

**Comment by wolfv** [comment]
argh, no idea what's going on. Might be network related. It's already pretty late over here so I won't be able to look into it more tonight. We also pushed a new rattler-build release which should at least speed things up and use fewer file handles. It might help, idk! if someone wants to shepherd that release now - it's already on github.

**Comment by h-vetinari** [comment]
> if someone wants to shepherd that release now - it's already on github.

Happy to help, though there's only a commit, no tag. Perhaps someone forget to push the tags?

**Comment by wolfv** [comment]
Just made it latest release and the tag should be there.

**Review by h-vetinari — APPROVED** [review]

**Comment by hadim** [comment]
Thanks both of you for the help here!

**Pull request merged**